content/documentation/3-stream-developer-guides/5-troubleshooting/1-debugging-apps.md

```diff
@@ -0,0 +1,182 @@
+---
+path: 'stream-developer-guides/troubleshooting/debugging-apps/'
+title: 'Debugging applications'
+description: 'Debugging applications'
+---
+
+Applications should be runnable as a standard Java JAR, invoked via `java -jar`.
+All dependencies such as databases and messaging middleware should be available and tested manually for connection issues.
+Running and debugging applications is independent of deploying via SCDF and running on a platform.
+
+# Project Build Errors
+
+```mermaid
+graph TD;
+    A(Errors in IDE) --> B(Run using Maven CLI);
+    B --> C{Success?}
+    C -->|Yes| D(Fix IDE errors)
+    C -->|No| E(Fix code errors)
+```
+
+Fixing errors in the IDE will be dependent on which IDE is used.
+By verifying the build is successful outside the IDE, this helps rule out project specific issues.
+
+# Application Startup Exceptions
+
+```mermaid
+graph TD;
+    A(Application Exceptions) --> B(Review stacktrace)
+    B --> C{General Coding Errors}
+    C --> D(Fix Code)
+    B --> E{Connection Exceptions}
+    E --> |Yes| F{Valid configuration?}
+    F --> |Yes| G(Test manually & apply any changes)
+    F --> |No| H(Fix settings)
+```
+
+When the application starts up, exceptions may occur due to general coding errors or connectivity issues.
+Review the application log file to find the exceptions and fix as appropriate.
+Ensure all external services are reachable, correct credentials are provided and any other required information.
+
+# General Application Debugging
+
+```mermaid
+graph TD;
+    A(Review stacktrace) --> B(Enable JDWP for debugging);
+    A(Review stacktrace) --> C(Enable Spring Integration debug logging);
+```
+
+Applications can be debugged as normal java processes as well as by utilizing verbose log output via the DEBUG logging level.
+
+To debug your application using a debugger, add the following JVM argument when starting your application, for example:
+
+`-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005`
+
+By adding this parameter, your application will wait for a debugger to connect on port `5005` to start debugging.
+This can be any port you choose that is not already in use.
+
+Enabling DEBUG level logging can also be a useful aid. One package of interest is `org.springframework.integration`.
+To enable debug logging of this package, add the following JVM argument when starting your application, for example:
+
+`-Dlogging.level.org.springframework.integration=DEBUG`
+
+# Data Loss
+
+```mermaid
+graph TD;
+    A(Data is lost) --> B{Middleware Accessible?}
+    B --> |Yes| C{Valid configuration?}
+    B --> |No| B
+    C --> |Yes| D(TestBinder in build test scope?)
+    D --> |Yes| G(Are messages in middleware?)
+    D --> |No| E(Move to test scope)
+    C --> |No| F(Test manually / fix settings)
+    G --> |No| H(Verify configuration)
+    H --> I(Verify Binder Configuration)
+    I --> J(Enable SI debug logging)
```

Comment:
Let's call out `Spring Integration = SI` somewhere in the guide.

Reply:
Maybe at line-58

Reply:
added